client/js/spotify-gdpr.js

```diff
@@ -0,0 +1,203 @@
+// @ts-check
+/* global c3 */
+
+// most recent ordering of top songs
+const latest = [
+  'song3',
+  'song2',
+  'song5',
+  'song7',
+  'song1',
+  'song10',
+  'song11',
+  'song12',
+  'song13',
+  'song14',
+  'song15',
+  'song16',
+  'song17',
+  'song18',
+  'song19',
+  'song20',
+  'song21',
+  'song22',
+  'song23',
+  'song24',
+  'song25',
+  'song26',
+  'song27',
+  'song28',
+  'song29',
+];
+
+/**
+ * @typedef {object} MoveChange
+ * @property {'move'} type
+ * @property {number} from
+ * @property {number} to
+ */
+
+/**
+ * @typedef {object} ReplaceChange
+ * @property {'replace'} type
+ * @property {string} oldItem
+ * @property {string} newItem
+ */
+
+/**
+ * @typedef {MoveChange | ReplaceChange} Change
+ */
+
+/**
+ * each array is a list of changes to the top songs list that happened on a
+ * given day
+ *
+ * @type {Change[][]}
+ */
+const changes = [
+  [{type: 'move', from: 2, to: 4}],
+  [{type: 'replace', oldItem: 'song4', newItem: 'song7'}],
+  [{type: 'move', from: 3, to: 4}],
+  [{type: 'move', from: 5, to: 4}],
+  [{type: 'move', from: 3, to: 4}],
+  [],
+  [],
+  [{type: 'replace', oldItem: 'song30', newItem: 'song23'}],
+  [{type: 'move', from: 6, to: 5}],
+  [{type: 'move', from: 7, to: 6}],
+  [{type: 'move', from: 9, to: 10}],
+  [{type: 'move', from: 20, to: 21}],
+  [],
+  [{type: 'move', from: 20, to: 21}],
+  [{type: 'move', from: 12, to: 13}],
+  [],
+  [],
+  [],
+];
+
+// number of days we have backtracked
+let generation = 1;
+const x = [new Date()];
+
+// current position of each song (list of songs in order)
+const current = latest;
+// current position of each song (map from song id to current index)
+const indices = new Map(current.map((id, idx) => [id, idx]));
+// historical position of each song (map from song id to array of historical
+// indices)
+const indicesHistory = new Map(current.map((id, idx) => [id, [idx]]));
+
+// for each changeset (set of changes that occurred on a day) apply the opposite
+// change so that we can reconstruct the ordering of the top songs list on that
+// day
+for (const changeSet of changes.reverse()) {
```

Comment:
`changes.reverse()` reverses the contents of the array in place. I suspect the two for loops here can be consolidated down into one:
```
const changeHistory = changes.slice().reverse();
for (const change of changeHistory) {
   ...
}
```

Reply:
That won't work because `changes` is an array of arrays. And it doesn't matter that the array is reversed in-place because I'm only iterating over it once, and the reversed array is returned.

Reply:
Ah, my mistake. Then you should either modify your original arrays on a separate line, or clone your arrays first in your loops:
```
changes.reverse();
for (changeSet of changes) {
  changeSet.reverse();
  for (change of changeSet) {
    ...
  }
}

// or
for (changeSet of changes.slice().reverse()) {
  for (change of changeSet.slice().reverse()) {
    ...
  }
}
}
```